### PR TITLE
feat: added the ability to enable/disable breakpoints and watchpoints

### PR DESCRIPTION
**Pull request by s4my**
This feature adds the ability to enable/disable `breakpoints` and `watchpoints` by giving the user the option to do so in the right click menu of the `Breakpoints` window, and the left margin of the code view.

The marker on the left margin of the code view of disabled breakpoints will be indicated in a darker shade of red `0x520404`, and for the `Breakpoints` window we have added a new column ("Enabled") that will indicate with yes or no whether the breakpoint/watchpoint is enabled or disabled.

### CODE REVIEW

**Comment by nakst** [comment]
Thanks for this pull request. Is there any way you can change it so that there aren't lots of whitespace modifications in the diff? It's a bit difficult to review at the moment.

**Comment by s4my** [comment]
> Is there any way you can change it so that there aren't lots of whitespace modifications in the diff? It's a bit difficult to review at the moment.

The code base is littered with trailing white spaces, which makes it annoying to see changes every time you save a file, so i created the 1st commit just to get rid of the trailing white spaces, and the 2nd one is the actual implementation of the feature in this commit 51126d8c4088175f7a7ce73949412ae61725ce36.

I hope in the futur we will add a code formatter like [Clang-Format](https://clang.llvm.org/docs/ClangFormat.html) to have a consistent formatting through out the code base.

File: gf2.cpp [hunk 1]
```diff
@@ -1003,6 +1016,22 @@ void CommandDeleteBreakpoint(void *_index) {
 	DebuggerSend(buffer, true, false);
 }
 
+void CommandDisableBreakpoint(void *_index) {
```
Comment (s4my) [inline gf2.cpp:1019]: We could merge CommandDisableBreakpoint() and CommandEnableBreakpoint() into a single function CommandToggleDisableBreakpoint(), but i don't know if keep them separate could comme in handy in the futur?

File: gf2.cpp [hunk 2]
```diff
@@ -1003,6 +1016,22 @@ void CommandDeleteBreakpoint(void *_index) {
...
+	DebuggerSend(buffer, true, false);
+}
+
+void CommandEnableBreakpoint(void *_index) {
+	int index = (int) (intptr_t) _index;
+	Breakpoint *breakpoint = &breakpoints[index];
+	char buffer[1024];
+	StringFormat(buffer, 1024, "enable %d", breakpoint->number);
+	DebuggerSend(buffer, true, false);
+}
```
Comment (s4my) [inline gf2.cpp:1033]: We could merge `CommandDisableBreakpoint()` and `CommandEnableBreakpoint()` to `CommandToggleDisableBreakpoint()`, but i don't know if keeping them separate could come in handy in the futur?

**Comment by nakst** [comment]
It seems not to handle well the case where you have two breakpoints on the same line where one is enabled and the other is disabled. But this is not too important for a first implementation.

**Comment by s4my** [comment]
I saw the commit 37b8ec1a5ef530218f77ad62a9e6d264987f6fab. you could of asked me to make those change (like use the ternary operator here, rename this var to this, change the color to this, don't make distinction between watchpoints and breakpoints,...etc) but i guess you eager to get going 😃 .

I don't know if you know but GitHub has this feature if you hover over a line number a blue bubble with + sign shows up, if clicked (or clicked and dragged) it would allow you to put comments in-place when you review code.

> It seems not to handle well the case where you have two breakpoints on the same line where one is enabled and the other is disabled. But this is not too important for a first implementation.

If i understand correctly. If for example i set a breakpoint on line twice from the console, since we only show the 1st one in the UI, disabling the 1st one should also disable the 2nd one.

**Comment by nakst** [comment]
> If i understand correctly. If for example i set a breakpoint on line twice from the console, since we only show the 1st one in the UI, disabling the 1st one should also disable the 2nd one, or even delete it all together.

I don't use enabling/disabling breakpoints in my workflow, so I don't know what changes could potentially need to be made. My advice is you should try using implementation as-is for now, and if any problems arise in your workflow because of this, make the most sensible change then :)

**Comment by s4my** [comment]
I implemented the fix here #122

**Pull request merged**